Allow webpushd to select web clip implicitly via origin

bnham opened the pull request:
#### fe1d144d508e27966b447ec6c7f400d84a63ce92
<pre>
Allow webpushd to select web clip implicitly via origin
<a href="https://bugs.webkit.org/show_bug.cgi?id=278768">https://bugs.webkit.org/show_bug.cgi?id=278768</a>
<a href="https://rdar.apple.com/134520632">rdar://134520632</a>

Reviewed by Brady Eidson.

For some webpushd clients, it makes more sense to allow clients to specify the web clip they want to
operate on implicitly via origin rather than explicitly passing in the web clip identifier. This can
make sense if e.g. a client wants multiple web clips to share the same service worker since they are
part of the same origin. In this case, we select the oldest web clip associated with the origin as
the representative web clip for that origin.

The way this works is that on iOS, we now allow webClipIdentifier/pushPartition to be empty at
connection startup time. If this is the case, then each IPC infers the effective webClipIdentifier.
This works by basically adding a subscriptionSetForOrigin call to every webpushd IPC.

Most IPCs already pass an origin, so this just works. I had to change a few IPCs to make them
compatible however:

 - cancelNotification now takes in an origin as well as a notification identifier (the origin is
   necessary for implicit web clip selection to work).
 - removeAllPushSubscriptions removes all subscriptions irrespective of web clip identifier when
   in implicit web clip mode. This required making some changes to PushDatabase and PushService.

I tested this by hooking up the push action handler in 282672@main to get and process push events.
However I left that part out of this patch since Brady is also working on that logic and want to
avoid merge conflicts. As part of that testing that hookup, I fixed some other issues:

 - The UIProcess-side permissions lookup in NetworkProcessProxy is not necessary anymore and I
   conditionalized it.
 - SBSCreateOpenApplicationService does not actually follow the create rule, which was causing
   webpushd crashes, so I stopped using that with adoptNS.
 - Hooked up MobileMiniBrowser&apos;s data store to its push action handler.

* Source/WebCore/Modules/push-api/PushDatabase.cpp:
(WebCore::PushDatabase::removeRecordsByBundleIdentifierAndDataStore):
* Source/WebCore/Modules/push-api/PushDatabase.h:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::cancelNotification):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::cancelNotification):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::cancelNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::cancel):
* Source/WebKit/webpushd/PushClientConnection.h:
(WebPushD::PushClientConnection::hostAppHasPushInjectEntitlement const):
(WebPushD::PushClientConnection::pushPartitionIfExists const):
(WebPushD::PushClientConnection::pushPartitionString const): Deleted.
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::isValidPushPartition):
(WebPushD::webClipIdentifierForOrigin):
(WebPushD::PushClientConnection::subscriptionSetIdentifierForOrigin const):
(WebPushD::PushClientConnection::debugDescription const):
(WebPushD::PushClientConnection::cancelNotification):
(WebPushD::PushClientConnection::subscriptionSetIdentifier const): Deleted.
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::removeRecordsForBundleIdentifierAndDataStore):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformNotificationCenterBundleIdentifier):
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::updateSubscriptionSetState):
(WebPushD::WebPushDaemon::setPushAndNotificationsEnabledForOrigin):
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::handleIncomingPushImpl):
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
(WebPushD::connectionMatchesPendingPushMessage):
(WebPushD::WebPushDaemon::getPendingPushMessage):
(WebPushD::WebPushDaemon::getPendingPushMessages):
(WebPushD::WebPushDaemon::subscribeToPushService):
(WebPushD::WebPushDaemon::unsubscribeFromPushService):
(WebPushD::WebPushDaemon::getPushSubscription):
(WebPushD::WebPushDaemon::incrementSilentPushCount):
(WebPushD::WebPushDaemon::removeAllPushSubscriptions):
(WebPushD::WebPushDaemon::removePushSubscriptionsForOrigin):
(WebPushD::WebPushDaemon::showNotification):
(WebPushD::WebPushDaemon::getNotifications):
(WebPushD::WebPushDaemon::cancelNotification):
(WebPushD::WebPushDaemon::getPushPermissionState):
(WebPushD::WebPushDaemon::requestPushPermission):
(WebPushD::WebPushDaemon::setAppBadge):
(WebPushD::WebPushDaemon::getAppBadgeForTesting):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/AppDelegate.m:
(-[AppDelegate application:didFinishLaunchingWithOptions:]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.h:
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController dataStore]):
(-[WebViewController createWebView]):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::PushDatabaseTest::removeRecordsByBundleIdentifierAndDataStore):
(TestWebKitAPI::TEST_F(PushDatabaseTest, RemoveRecordsByBundleIdentifierAndDataStore)):

Canonical link: <a href="https://commits.webkit.org/282857@main">https://commits.webkit.org/282857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f380f687a4e797eac77d22751c00e9a9fe2b5fc4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64495 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17091 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68518 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66614 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15382 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/10421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55813 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/32518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/37236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13190 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13976 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/13519 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70217 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8442 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8476 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55901 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14223 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6984 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/39673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40494 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
```diff
@@ -43,6 +43,7 @@
 
 @interface BSActionResponse : NSObject
 + (instancetype)response;
++ (instancetype)responseForError:(NSError *)error;
```
I don't think this is used anywhere in the patch?

beidson commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -673,50 +660,70 @@ static void tryCloseRequestConnection(xpc_object_t request)
+    auto makeNotNull = [](String s) {
+        return s.isNull() ? emptyString() : s;
+    };
+    return makeNotNull(connection.pushPartitionIfExists()) == makeNotNull(identifierForPendingPushMessage.pushPartition);
```
equalIgnoringNullity

bnham committed:
Allow webpushd to select web clip implicitly via origin
https://bugs.webkit.org/show_bug.cgi?id=278768
rdar://134520632

Reviewed by Brady Eidson.

For some webpushd clients, it makes more sense to allow clients to specify the web clip they want to
operate on implicitly via origin rather than explicitly passing in the web clip identifier. This can
make sense if e.g. a client wants multiple web clips to share the same service worker since they are
part of the same origin. In this case, we select the oldest web clip associated with the origin as
the representative web clip for that origin.

The way this works is that on iOS, we now allow webClipIdentifier/pushPartition to be empty at
connection startup time. If this is the case, then each IPC infers the effective webClipIdentifier.
This works by basically adding a subscriptionSetForOrigin call to every webpushd IPC.

Most IPCs already pass an origin, so this just works. I had to change a few IPCs to make them
compatible however:

 - cancelNotification now takes in an origin as well as a notification identifier (the origin is
   necessary for implicit web clip selection to work).
 - removeAllPushSubscriptions removes all subscriptions irrespective of web clip identifier when
   in implicit web clip mode. This required making some changes to PushDatabase and PushService.

I tested this by hooking up the push action handler in 282672@main to get and process push events.
However I left that part out of this patch since Brady is also working on that logic and want to
avoid merge conflicts. As part of that testing that hookup, I fixed some other issues:

 - The UIProcess-side permissions lookup in NetworkProcessProxy is not necessary anymore and I
   conditionalized it.
 - SBSCreateOpenApplicationService does not actually follow the create rule, which was causing
   webpushd crashes, so I stopped using that with adoptNS.
 - Hooked up MobileMiniBrowser's data store to its push action handler.

* Source/WebCore/Modules/push-api/PushDatabase.cpp:
(WebCore::PushDatabase::removeRecordsByBundleIdentifierAndDataStore):
* Source/WebCore/Modules/push-api/PushDatabase.h:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::cancelNotification):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::cancelNotification):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::cancelNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::cancel):
* Source/WebKit/webpushd/PushClientConnection.h:
(WebPushD::PushClientConnection::hostAppHasPushInjectEntitlement const):
(WebPushD::PushClientConnection::pushPartitionIfExists const):
(WebPushD::PushClientConnection::pushPartitionString const): Deleted.
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::isValidPushPartition):
(WebPushD::webClipIdentifierForOrigin):
(WebPushD::PushClientConnection::subscriptionSetIdentifierForOrigin const):
(WebPushD::PushClientConnection::debugDescription const):
(WebPushD::PushClientConnection::cancelNotification):
(WebPushD::PushClientConnection::subscriptionSetIdentifier const): Deleted.
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::removeRecordsForBundleIdentifierAndDataStore):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformNotificationCenterBundleIdentifier):
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::updateSubscriptionSetState):
(WebPushD::WebPushDaemon::setPushAndNotificationsEnabledForOrigin):
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::handleIncomingPushImpl):
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
(WebPushD::connectionMatchesPendingPushMessage):
(WebPushD::WebPushDaemon::getPendingPushMessage):
(WebPushD::WebPushDaemon::getPendingPushMessages):
(WebPushD::WebPushDaemon::subscribeToPushService):
(WebPushD::WebPushDaemon::unsubscribeFromPushService):
(WebPushD::WebPushDaemon::getPushSubscription):
(WebPushD::WebPushDaemon::incrementSilentPushCount):
(WebPushD::WebPushDaemon::removeAllPushSubscriptions):
(WebPushD::WebPushDaemon::removePushSubscriptionsForOrigin):
(WebPushD::WebPushDaemon::showNotification):
(WebPushD::WebPushDaemon::getNotifications):
(WebPushD::WebPushDaemon::cancelNotification):
(WebPushD::WebPushDaemon::getPushPermissionState):
(WebPushD::WebPushDaemon::requestPushPermission):
(WebPushD::WebPushDaemon::setAppBadge):
(WebPushD::WebPushDaemon::getAppBadgeForTesting):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/AppDelegate.m:
(-[AppDelegate application:didFinishLaunchingWithOptions:]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.h:
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController dataStore]):
(-[WebViewController createWebView]):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::PushDatabaseTest::removeRecordsByBundleIdentifierAndDataStore):
(TestWebKitAPI::TEST_F(PushDatabaseTest, RemoveRecordsByBundleIdentifierAndDataStore)):

Canonical link: https://commits.webkit.org/282857@main
